### PR TITLE
Convert StateService to suspend functions

### PR DESCRIPTION
**Pull request by aqulu**
Signed-off-by: aqulu <dev@aqu.lu>

### Pull Request Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [x] Pull request is based on the develop branch
- [x] Pull request updates [CHANGES.md](https://github.com/vector-im/element-android/blob/develop/CHANGES.md)
- [ ] Pull request includes screenshots or videos if containing UI changes
- [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)

### Summary

- Converted StateService to suspendable functions #2449
- RxRoom launches suspendable functions using `kotlinx-coroutines-rx2` rxCompletable
    - rxCompletable internally launches the suspendable function on `GlobalScope`. The job will be cancelled when the resulting Completable is, which is similar to `completableBuilder`'s behavior in matrix-sdk-android-rx
- WidgetPostAPIHandler uses `GlobalScope.launch` to run suspendable functions
    - __Attention:__ Directly using GlobalScope is discouraged.
    - As WidgetPostAPIHandler is not tied to any lifecycle. Using a new `CoroutineScope` instance does not solve the issue, as running jobs would not be cancelled by lifecycle events.
        - _WidgetPostAPIHandler seems to be used only by `WidgetViewModel` so it could potentially be tied to it's lifecycle to avoid using GlobalScope._
    - Prior to these changes, the function calls were not constrained either, thus using GlobalScope seems appropriate.
        - StateService internally used `TaskExecutor`'s coroutine scope. All jobs in this scope will be cancelled when CacheService.clearCache is invoked. As far as i understand this is a side-effect of cache clearing rather than it's primary goal.

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt [hunk 1]
```diff
@@ -73,45 +67,41 @@ internal class DefaultStateService @AssistedInject constructor(@Assisted private
...
-                body = body
-        )
-        return sendStateTask
-                .configureWith(params) {
-                    this.callback = callback
-                }
-                .executeBy(taskExecutor)
+            body: JsonDict
+    ) {
+        withContext(coroutineDispatchers.main) {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt:75]: Why changing the context here?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt [hunk 2]
```diff
@@ -123,64 +113,52 @@ internal class DefaultStateService @AssistedInject constructor(@Assisted private
...
                 body = mapOf("history_visibility" to readability),
-                callback = callback,
                 stateKey = null
         )
     }
 
-    override fun updateJoinRule(joinRules: RoomJoinRules?, guestAccess: GuestAccess?, callback: MatrixCallback<Unit>): Cancelable {
-        return taskExecutor.executorScope.launchToCallback(coroutineDispatchers.main, callback) {
+    override suspend fun updateJoinRule(joinRules: RoomJoinRules?, guestAccess: GuestAccess?) {
+        withContext(coroutineDispatchers.main) {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt:129]: same question here

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt [hunk 3]
```diff
@@ -123,64 +113,52 @@ internal class DefaultStateService @AssistedInject constructor(@Assisted private
...
+                        stateKey = null
+                )
             }
         }
     }
 
-    override fun updateAvatar(avatarUri: Uri, fileName: String, callback: MatrixCallback<Unit>): Cancelable {
-        return taskExecutor.executorScope.launchToCallback(coroutineDispatchers.main, callback) {
+    override suspend fun updateAvatar(avatarUri: Uri, fileName: String) {
+        withContext(coroutineDispatchers.main) {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt:148]: and here

**Review by bmarty — COMMENTED** [review]
Thanks for your PR.
I got just one question, and I will be more comfortable if @ganfra can have a look on the change regarding the Widget part.
Also the CI is not happy, could you fix that please?
Thanks!

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt (hunk 1 above)
Comment (aqulu) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt:75]: Sorry, didn't read `executeBy(taskExecutor)` internals properly.
TaskExecutor.execute ran this on the callback's thread (by transforming it with `toDispatcher()` internally). I'll remove the context change.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt (hunk 2 above)
Comment (aqulu) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt:129]: This was launched this in the `coroutineDispatchers.main` context by `taskExecutor.executorScope.launchToCallback(coroutineDispatchers.main, callback)`  before.
I'm happy to remove the explicit `withContext` though.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt (hunk 3 above)
Comment (aqulu) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/state/DefaultStateService.kt:148]: same as https://github.com/vector-im/element-android/pull/2500/files/c7efd1feb99179585e865affa866d680a96fefe2#r538291540

**Comment by aqulu** [comment]
Thanks for the review!

>  Also the CI is not happy, could you fix that please?

I'll fix the CI errors and get rid of the context-changes you pointed out.

**Comment by aqulu** [comment]
CI should be fixed now. Looks like buildkite's "Compile and run Unit tests" job hung up - executing the command on my local machine completes successfully.

File: vector/src/main/java/im/vector/app/features/widgets/WidgetPostAPIHandler.kt [hunk 4]
```diff
@@ -310,12 +312,19 @@ class WidgetPostAPIHandler @AssistedInject constructor(@Assisted private val roo
...
-        )
+
+        GlobalScope.launch {
+            try {
+                room.sendStateEvent(
+                        eventType = EventType.PLUMBING,
+                        stateKey = null,
+                        body = params
+                )
+                widgetPostAPIMediator.sendSuccess(eventData)
```
Comment (Dominaezzz) [inline vector/src/main/java/im/vector/app/features/widgets/WidgetPostAPIHandler.kt:323]: Should this be done inside the `try` block? If it throws, `sendError` will also be called, which isn't strictly the previous behaviour.
(Same for below)

**Comment by Dominaezzz** [comment]
I've been quite busy lately and haven't been able to contribute to this, but I wasn't expecting anybody else to do these.
Might be worth "reserving" each service in the tracking issue so no one duplicates work because these can be annoying enough to untangle and maintain previous behaviour, don't want my (or anyone else's) work to be made redundant :sob:.

Also, feel free to tag me to review these.

File: vector/src/main/java/im/vector/app/features/widgets/WidgetPostAPIHandler.kt (hunk 4 above)
Comment (aqulu) [reply vector/src/main/java/im/vector/app/features/widgets/WidgetPostAPIHandler.kt]: Thanks for pointing this out! Fixed in https://github.com/vector-im/element-android/pull/2500/commits/ed822becc68c16e03330b0139a7287861ccf72de

**Comment by aqulu** [comment]
> I've been quite busy lately and haven't been able to contribute to this, but I wasn't expecting anybody else to do these.

@Dominaezzz Sorry for interfering with these, that wasn't my intention :bow:
I'll hold off from the other converting tasks, as it might be more consistent if done by the same person.

**Comment by Dominaezzz** [comment]
Oh no! :fearful: Don't stop on my behalf! It would be amazing to have these done sooner than later aha.
I don't currently have a lot of time, so it would be awesome if you continued.
Consistency can be solved with reviews but I don't think there's much to worry about there anyway.

Looking forward to your next PR! :tada: :grinning:

**Review by bmarty — APPROVED** [review]
LGTM now, thanks for the update!
It seems that we have touch the StateService. Would you mind to rebase your change on develop, or merge develop to your branch?

**Comment by bmarty** [comment]
Actually I think I can do it with Github UI

**Pull request merged**